File: backend/models/survey.py
# Created by Ryan Polasky | 7/12/25
# ACM MeteorMate | All Rights Reserved

import enum

from sqlalchemy import (
    Boolean,
    Column,
    Date,
    DateTime,
    ForeignKey,
    Integer,
    Text,
    text,
)
from sqlalchemy import Enum as SAEnum
from sqlalchemy.dialects.postgresql import ARRAY, JSONB
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func

from ..database import ORMBase


def mm_enum(enum_cls: type[enum.Enum], name: str) -> SAEnum:
    return SAEnum(enum_cls, name=name, values_callable=lambda e: [x.value for x in e])


# Shared housing enums
class HousingIntentEnum(str, enum.Enum):
    ON_CAMPUS = "on_campus"
    OFF_CAMPUS = "off_campus"
    BOTH = "both"


# Wake/Clean/Noise tile enums
class WakeTimeEnum(str, enum.Enum):
    EARLY_BIRD = "early_bird"
    FLEXIBLE = "flexible"
    NIGHT_OWL = "night_owl"


class CleanlinessEnum(str, enum.Enum):
    RELAXED = "relaxed"
    TIDY = "tidy"
    NEAT_FREAK = "neat_freak"


class NoiseToleranceEnum(str, enum.Enum):
    QUIET = "quiet"
    MODERATE = "moderate"
    LOUD = "loud"


# Lifestyle tile enums
class CookingFrequencyEnum(str, enum.Enum):
    NEVER = "never"
    RARELY = "rarely"
    OFTEN = "often"


class PetPreferenceEnum(str, enum.Enum):
    OKAY = "okay"
    NOT_OKAY = "not_okay"
    HAVE_A_PET = "have_a_pet"


class GuestsFrequencyEnum(str, enum.Enum):
    NEVER = "never"
    SOMETIMES = "sometimes"
    OFTEN = "often"


class RoommateClosenessEnum(str, enum.Enum):
    NOT_CLOSE = "not_close"
    FRIENDS = "friends"
    CLOSE_FRIENDS = "close_friends"


# Dealbreakers
class DealbreakerEnum(str, enum.Enum):
    SMOKE_VAPE = "smoke_vape"
    DRINK = "drink"
    SAME_GENDER = "same_gender"


# On-campus flow
class OnCampusLocationEnum(str, enum.Enum):
    UV = "uv"
    CC = "cc"
    FRESHMAN_DORMS = "freshman_dorms"
    NORTHSIDE = "northside"


class NumRoommatesEnum(str, enum.Enum):
    NO_PREFERENCE = "no_preference"
    ONE = "one"
    TWO = "two"
    THREE = "three"


# Off-campus "have a lease" flow
class HaveLeaseLengthEnum(str, enum.Enum):
    SEMESTER = "semester"
    ACADEMIC_YEAR = "academic_year"
    YEAR = "year"


class Survey(ORMBase):
    __tablename__ = "surveys"

    user_id = Column(
        Text,
        ForeignKey("users.id", ondelete="CASCADE"),
        primary_key=True,
        index=True,
    )

    # Core housing fields
    housing_intent = Column(mm_enum(HousingIntentEnum, "housing_intent_enum"), nullable=True)

    # Sliders (rent/budget)
    budget_min = Column(Integer, nullable=True)
    budget_max = Column(Integer, nullable=True)

    move_in_date = Column(Date, nullable=True)

    # Wake/Clean/Noise
    wake_time = Column(mm_enum(WakeTimeEnum, "wake_time_enum"), nullable=True)
    cleanliness = Column(mm_enum(CleanlinessEnum, "cleanliness_enum"), nullable=True)
    noise_tolerance = Column(mm_enum(NoiseToleranceEnum, "noise_tolerance_enum"), nullable=True)

    # Interests
    interests = Column(ARRAY(Text), nullable=False, server_default="{}")

    # Dealbreakers
    dealbreakers = Column(
        ARRAY(mm_enum(DealbreakerEnum, "dealbreaker_enum")),
        nullable=False,
        server_default="{}",
    )

    # Lifestyle / Personality
    cooking_frequency = Column(
        mm_enum(CookingFrequencyEnum, "cooking_frequency_enum"), nullable=True
    )
    pet_preference = Column(mm_enum(PetPreferenceEnum, "pet_preference_enum"), nullable=True)
    guests_frequency = Column(mm_enum(GuestsFrequencyEnum, "guests_frequency_enum"), nullable=True)
    roommate_closeness = Column(
        mm_enum(RoommateClosenessEnum, "roommate_closeness_enum"), nullable=True
    )

    # On-campus flow
    on_campus_locations = Column(
        ARRAY(mm_enum(OnCampusLocationEnum, "on_campus_location_enum")),
        nullable=False,
        server_default="{}",
    )

    # asked on on-campus page
    honors = Column(Boolean, nullable=True)

    llc_interest = Column(Boolean, nullable=True)

    # UV/CC/Northside stuff
    num_roommates = Column(mm_enum(NumRoommatesEnum, "num_roommates_enum"), nullable=True)

    # Off-campus lease flow
    have_lease = Column(Boolean, nullable=True)
    have_lease_length = Column(
        mm_enum(HaveLeaseLengthEnum, "have_lease_length_enum"), nullable=True
    )

    # Catch-all
    answers = Column(
        JSONB,
        nullable=False,
        server_default=text("'{}'::jsonb"),
    )

    created_at = Column(DateTime(timezone=True), nullable=False, server_default=func.now())
    updated_at = Column(
        DateTime(timezone=True),
        nullable=False,
        server_default=func.now(),
        onupdate=func.now(),
    )

    smoke_vape = Column(Boolean, nullable=False, server_default=text("false"))
    drink = Column(Boolean, nullable=False, server_default=text("false"))

    user = relationship("User", back_populates="survey", uselist=False)

File: src/types/survey.ts
export type HousingIntent = "on_campus" | "off_campus" | "both";

export type WakeTime = "early_bird" | "flexible" | "night_owl";

export type Cleanliness = "relaxed" | "tidy" | "neat_freak";

export type NoiseTolerance = "quiet" | "moderate" | "loud";

export type CookingFrequency = "never" | "rarely" | "often";

export type PetPreference = "okay" | "not_okay" | "have_a_pet";

export type GuestsFrequency = "never" | "sometimes" | "often";

export type RoommateCloseness = "not_close" | "friends" | "close_friends";

export type Dealbreaker = "smoke_vape" | "drink" | "same_gender";

export type OnCampusLocation = "uv" | "cc" | "freshman_dorms" | "northside";

export type NumRoommates = "no_preference" | "one" | "two" | "three";

export type HaveLeaseLength = "semester" | "academic_year" | "year";

export interface SurveyCreateBody {
    housing_intent?: HousingIntent | null;

    budget_min?: number | null;
    budget_max?: number | null;
    move_in_date?: string | null;

    wake_time?: WakeTime | null;
    cleanliness?: Cleanliness | null;
    noise_tolerance?: NoiseTolerance | null;

    interests?: string[];
    dealbreakers?: Dealbreaker[];

    cooking_frequency?: CookingFrequency | null;
    pet_preference?: PetPreference | null;
    guests_frequency?: GuestsFrequency | null;
    roommate_closeness?: RoommateCloseness | null;

    on_campus_locations?: OnCampusLocation[];
    honors?: boolean | null;
    llc_interest?: boolean | null;
    num_roommates?: NumRoommates | null;

    have_lease?: boolean | null;
    have_lease_length: HaveLeaseLength;

    answers?: Record<string, unknown>;

    smoke_vape?: boolean;
    drink?: boolean;
}

export interface SurveyUpdateBody {
    housing_intent?: HousingIntent | null;

    budget_min?: number | null;
    budget_max?: number | null;
    move_in_date?: string | null;

    wake_time?: WakeTime | null;
    cleanliness?: Cleanliness | null;
    noise_tolerance?: NoiseTolerance | null;

    interests?: string[] | null;
    dealbreakers?: Dealbreaker[] | null;

    cooking_frequency?: CookingFrequency | null;
    pet_preference?: PetPreference | null;
    guests_frequency?: GuestsFrequency | null;
    roommate_closeness?: RoommateCloseness | null;

    on_campus_locations?: OnCampusLocation[] | null;
    honors?: boolean | null;
    llc_interest?: boolean | null;
    num_roommates?: NumRoommates | null;

    have_lease?: boolean | null;
    have_lease_length?: HaveLeaseLength | null;

    answers?: Record<string, unknown> | null;

    smoke_vape?: boolean;
    drink?: boolean;
}

// Shared payload used by upsert flows (POST first, then PUT fallback)
export interface SurveyPayload {
    housing_intent?: string | null;

    budget_min?: number | null;
    budget_max?: number | null;
    move_in_date?: string | null;

    wake_time?: string | null;
    cleanliness?: string | null;
    noise_tolerance?: string | null;

    interests?: string[];
    dealbreakers?: string[];

    cooking_frequency?: string | null;
    pet_preference?: string | null;
    guests_frequency?: string | null;
    roommate_closeness?: string | null;

    on_campus_locations?: string[];
    honors?: boolean | null;
    llc_interest?: boolean | null;
    num_roommates?: string | null;

    have_lease?: boolean | null;
    have_lease_length?: string | null;

    answers?: Record<string, unknown>;

    smoke_vape?: boolean;
    drink?: boolean;
}

export interface SurveyResponse {
    user_id: string;

    housing_intent?: HousingIntent | null;

    budget_min?: number | null;
    budget_max?: number | null;
    move_in_date?: string | null;

    wake_time?: WakeTime | null;
    cleanliness?: Cleanliness | null;
    noise_tolerance?: NoiseTolerance | null;

    interests: string[];
    dealbreakers: Dealbreaker[];

    cooking_frequency?: CookingFrequency | null;
    pet_preference?: PetPreference | null;
    guests_frequency?: GuestsFrequency | null;
    roommate_closeness?: RoommateCloseness | null;

    on_campus_locations: OnCampusLocation[];
    honors?: boolean | null;
    llc_interest?: boolean | null;
    num_roommates?: NumRoommates | null;

    have_lease?: boolean | null;
    have_lease_length: HaveLeaseLength;

    answers: Record<string, unknown>;

    created_at: string;
    updated_at: string;

    smoke_vape: boolean;
    drink: boolean;
}